symengine/python/symengine/lib/symengine_wrapper.pyx

```diff
@@ -1702,5 +1702,313 @@ def eval_double(basic):
+    cdef size_t i
+    for i in range(vec.size()):
+        out[i] = symengine.eval_double(deref(
+            <symengine.RCP[const symengine.Basic]>(vec[i])))
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef void as_complex(symengine.vec_basic vec, double complex[::1] out) nogil:
+    cdef size_t i
+    for i in range(vec.size()):
+        out[i] = symengine.eval_complex_double(deref(
+            <symengine.RCP[const symengine.Basic]>(vec[i])))
+
+
+cdef size_t _size(n):
+    try:
+        return n.size
+    except AttributeError:
+        return len(n)  # e.g. array.array
+
+def get_buffer_address(arr):
+    try:
+        return arr.data
+    except AttributeError:
+        return arr.buffer_info()[0]  # np.array
+
+def with_buffer(iterable):
+    """ if iterable supports the buffer interface: return iterable,
+        if not, return a cython.view.array object (which does) """
+    cdef double[::1] cy_arr_view
+    try:
+        if PY_MAJOR_VERSION == 2 and PY_MINOR_VERSION == 6:
+            eval('buffer(iterable)')
+        else:
+            memoryview(iterable)
+        # When dropping 2.6, above may be replaced by:
+        # memoryview(iterable)
+    except TypeError:
+        # Cython's array to the rescue
+        cy_arr =  cython.view.array(shape=(_size(iterable),),
+                                    itemsize=sizeof(double), format='d')
+        cy_arr_view = cy_arr
+        for i in range(_size(iterable)):
+            cy_arr_view[i] = iterable[i]
+        return cy_arr
+    else:
+        return iterable  # iterable supports memoryview
+
+
+cdef class Lambdify(object):
+    """
+    Parameters
+    ----------
+    args: iterable of Symbols
+    exprs: array_like of expressions
+
+    Returns
+    -------
+    callback instance with signature f(inp, out=None)
+
+    Examples
+    --------
+    >>> from symengine import var, Lambdify
+    >>> var('x y z')
+    >>> f = Lambdify([x, y, z], [x+y+z, x*y*z])
+    >>> f([2, 3, 4])
+    [ 9., 24.]
+    >>> out = np.array(2)
+    >>> f(x, out); out
+    [ 9., 24.]
+
+    """
+    cdef size_t in_size, out_size
+    cdef tuple out_shape
+    cdef symengine.vec_basic args, exprs
+
+    def __cinit__(self, args, exprs):
+        cdef Basic e_
+        try:
+            self.out_shape = exprs.shape
+        except AttributeError:
+            self.out_shape = len(exprs),
+        self.in_size = len(args)
+        self.out_size = reduce(mul, self.out_shape)
+
+        for e in args:
+            e_ = sympify(e)
+            self.args.push_back(e_.thisptr)
+
+        try:
+            exprs = exprs.flatten()
+        except AttributeError:
+            pass
+
+        for e in exprs:
+            e_ = sympify(e)
+            self.exprs.push_back(e_.thisptr)
+
+    # Blow follow 4 versions of same code double/complex input/output:
+    # we could replace this with some simple code generation, but for
+    # now it is easier to manually keep these 4 in sync.
```

Comment:
Duplication can be avoided using Cython's `fused_type`s
http://docs.cython.org/src/userguide/fusedtypes.html